homeassistant/components/homekit/type_fans.py

```diff
@@ -41,19 +46,29 @@ def __init__(self, *args):
             chars.append(CHAR_ROTATION_DIRECTION)
         if features & SUPPORT_OSCILLATE:
             chars.append(CHAR_SWING_MODE)
+        if features & SUPPORT_SET_SPEED:
+            chars.append(CHAR_ROTATION_SPEED)
```

Comment:
Before adding the `char`, we would need to make sure we support the specific speed list.
Maybe extract that check into a new method, something along the lines of:
```py
def check_speed_list(self):
  speed_list = self.hass.states.get(self.entity_id).attributes.get(ATTR_SPEED_LIST)
  if speed_list == [SPEED_OFF, SPEED_LOW, SPEED_MEDIUM, SPEED_HIGH]:
    return True
  return False
```

```py
if features & SUPPORT_SET_SPEED and self.check_speed_list():
    chars.append(CHAR_ROTATION_SPEED)
```

Reply:
As an alternative `check_speed_list` could return an integer we use for internal mapping: `speed_list_type`.

Reply:
As long as speed_list contains arbitrary values, I cannot think of a way of using and integer. **Only if** we are restricting the list to the base modes I imagine that I can use a constant indicating the possibilities _speed list with off, low, high_ and _speed list with off, low, medium, high_, but this would be very brittle and break even if something trivial like a fourth speed suggested in home-assistant/architecture#127 gets introduced.

I think a better idea something that contains a mapping between Home Assistant speed values and percentages; the percentages must allow for ranges since a fan with `off`, `low`, `high` should behave like this to be intuitive (IMHO):

- 0%-33% should map to `off`
- `off` should be shown as 0% (otherwise the user as the strange experience that the fan is off, but still shows a non-zero value)
- 34%-66% should map to `low`
- `low` should be shown as 50%
- 67%-100% should map to `high`
- `high` should be shown as 100%

The bounds of the ranges can be set either implicitly or explicitly. If everything is implicit, then this would be equivalent to storing `speed_list`, just as I did in this PR.

I think the best way to do this is to use list/set containing `namedtuple("PercentageMapping", "speed_value percentage_target percentage_start percentage_stop")` for the `speed_list_mapping`.

Another way is to use a dict: `speed_value_to_percentage` and a sorted list containing tuples(`percentage_stop`, `speed_value`), but in my opinion the first option is easier to understand, read, and debug.

Reply:
Please ignore the architecture issues for now.

As for the mapping, I agree. Sonething like:
Speed -> Target, Start, Stop
does make sense.